README
This is netkit-ftp-0.10.

This package updates the ftp client from NetKit-0.09. NetKit as a
package no longer exists. To do a full update of NetKit-0.09 please
consult NetKit.txt from your local ftp archive.

Contents:
	This package contains the ftp client program, ftp.

	If you're reading this off a CD, go right away and check the
	net archives for later versions and security fixes.
	
Requirements: (I know these work; you can probably use much earlier 
versions of libc, gcc, or the kernel.)

	libc 5.4 or higher 
		Older libc versions will work but anything before 5.4.6 
		includes some nasty security bugs. All the NetKit programs 
		are believed immune to these, but the rest of your system 
		probably isn't... 

		Note: libc 5.0.x is not supported and won't work.
		Note: glibc (aka libc 6.x) may or may not work with
		this release.

	linux 2.0 or higher 
		Use of 1.2.x kernel versions is not recommended in general.
		NetKit itself will probably work with anything since 0.99pl14,
		although some hacking would probably be required.

	gcc 2.7.2 or higher
		Please get the strength-reduce fix if you don't have it.

	libncurses
		Any recent version should be fine.

	libreadline (GNU readline library) is optional but recommended.

Please make sure your header files in /usr/include match your libc
version installed in /lib and /usr/lib. If you have weird problems
this is the most likely culprit.

Installation:
	You may need to patch your readline library. See below.
	Do "./configure", or edit MCONFIG to set the configuration options. 
	Do "make".
	Do "make install" as root to install everything. Save backup copies 
	  of important tools, or confirm the new ones work first. Etc. We 
	  warned you.

Other notes:
	To compile ftp with readline support you'll need a copy of the
	readline library installed. If you don't have it, get
	readline-2.0 from prep.ai.mit.edu (or any GNU mirror). You'll
	need to apply the enclosed patch, or ftp will have problems with 
	^C handling.

Security:
	This release contains no security fixes relative to the ftp
	client from NetKit-0.09. Versions prior to NetKit-0.09 should
	not be used.

Bugs:
	If something doesn't compile for you, send diffs. If you can't
	send diffs, send the compiler's error output.

	If it compiles but doesn't work, send as complete a bug report as 
	you can. Patches and fixes are welcome, as long as you describe 
	adequately what they're supposed to fix. Please, one patch per
	distinct fix.

	Be sure to send all correspondence in e-mail. Postings to netnews 
	will not be seen due to the enormous volume.

	Please don't report known bugs (see the BUGS file(s)) unless you
	are including fixes. :-)


David A. Holland	dholland@hcs.harvard.edu
								8-Mar-1997


ChangeLog
08-Jun-1997:
	Fixed doc for "newer" command. (Roderick Schertler,
	  roderick@argon.org) 

05-Apr-1997:
	Added configure script to generate MCONFIG.
	glibc fixes from HJ Lu.

20-Mar-1997:
	If using readline, pass it the terminal type, so certain .inputrc
	  constructions work. (Christian Groessler, chris@fast-ag.de)

08-Mar-1997: 
	Split from full NetKit package. 
	Generated this change log from NetKit's.

07-Mar-1997
	Changed order of includes in ftp/cmds.c due to a report of
	  breakage with libc 5.4.x.
	Fixed crash in ftp caused by certain syntax errors in .netrc.
	Fix symbol name conflict on "trace" between ftp and ncurses.

29-Dec-1996
	NetKit-0.09 released.
	Assorted alpha/glibc patches. (Erik Troan, ewt@redhat.com)
	Assorted bug fixes from Debian. (Peter Tobias, 
	  tobias@et-inf.fho-emden.de)
	Hardened programs against DNS h_length spoofing attacks.
	Use inet_aton() everywhere instead of inet_addr().
	Fixed crash in ftp if you did "lcd" with no args.
	Fixed bug in ftp where you couldn't abort an mget.
	Fixed /tmp security problem in ftp.

22-Aug-1996
	NetKit-B-0.08 released.
	(almost) everything now compiles with lots of warnings turned on.
	Fixed some memory allocation bugs in ftp.

25-Jul-1996
	NetKit-B-0.07A released.

23-Jul-1996
	NetKit-B-0.07 released.
	Integrated a collection of patches that had been lurking on the net,
	  including the 256-ptys support for telnetd and passive mode ftp.
	Major security fixes, including to fingerd, lpr, rlogin, rsh, talkd, 
	  and telnetd. Do *not* use the sliplogin from earlier versions of this
	  package, either.
	Much of the code builds without libbsd.a or bsd includes.
	Massive code cleanup. Almost everything compiles clean with gcc
	  -Wall now. rusers and rusersd do not; patches to rpcgen to fix
	  this would be appreciated if anyone feels like it.
	New maintainer:  David A. Holland, dholland@hcs.harvard.edu

date not known
	NetKit-B-0.06 released.

date not known
	NetKit-B-0.05 released.

date not known
	NetKit-B-0.04 released.

date not known
	NetKit-B-0.03 released.



ftp/glob.c
/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * From: @(#)glob.c	5.9 (Berkeley) 2/25/91
 */
char glob_rcsid[] = 
  "$Id: glob.c,v 1.5 1996/08/14 23:27:28 dholland Exp $";

/*
 * C-shell glob for random programs.
 */

#include <sys/param.h>
#include <sys/stat.h>
#include <dirent.h>

#include <pwd.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "ftp_var.h"  /* for protos only */
#include "glob.h"

#define	QUOTE 0200
#define	TRIM 0177
#define	eq(a,b)		(strcmp(a, b)==0)
#define	GAVSIZ		(ARG_MAX/6)
#define	isdir(d)	((d.st_mode & S_IFMT) == S_IFDIR)

const char *globerr;
extern const char *home;

typedef struct {
    const char *text;
} centry;

typedef struct {
    char *text;
} entry;

static entry *gargv;	/* Pointer to the (stack) arglist */
static entry *sortbas;
static int gargc;	/* Number args in gargv */
static int gnleft;      /* space left before we hit max args length */

static short gflag;

static int globcnt;
static const char *globchars = "`{[*?";

static char *gpath;
static char *gpathp, *lastgpathp;
static int globbed;
static const char *entp;

static int tglob(char c);
static char *strspl(const char *, const char *);
static char *strend(char *);
static char **copyblk(entry *);
static char **cloneblk(const centry *);

static void addpath(char c);
static void acollect(const char *);
static void collect(const char *);
static void expand(const char *);
static void Gcat(const char *, const char *);
static void ginit(entry *);
static void matchdir(const char *);
static void rscan(centry *, int (*f)(char));
static void sort(void);
static void efree(entry *);
static int amatch(const char *, const char *);
static int execbrc(const char *, const char *);
static int match(const char *, const char *);

static int gethdir(char *homedir);
static int letter(char c);
static int digit(char c);
static int any(int c, const char *s);

char **
ftpglob(const char *v)
{
	char agpath[BUFSIZ];
	entry agargv[GAVSIZ];
	centry vv[2];
	vv[0].text = v;
	vv[1].text = NULL;
	gflag = 0;
	rscan(vv, tglob);
	if (gflag == 0) {
		return cloneblk(vv);
	}

	globerr = 0;

	gpath = agpath; 
	gpathp = gpath; 
	*gpathp = 0;
	lastgpathp = agpath + sizeof(agpath) - 2;

	ginit(agargv); 
	globcnt = 0;
	collect(v);
	if (globcnt == 0 && (gflag&1)) {
		efree(gargv);
		gargv = NULL;
		return NULL;
	} 
	else {
		char **rv = copyblk(gargv);
		gargv = NULL;
		return rv;
	}
}

static 
void
ginit(entry *agargv)
{
	agargv[0].text = NULL; 
	gargv = agargv; 
	sortbas = agargv; 
	gargc = 0;
	gnleft = ARG_MAX - 4;
}

static 
void
collect(const char *as)
{
	if (eq(as, "{") || eq(as, "{}")) {
		Gcat(as, "");
		sort();
	} 
	else {
		acollect(as);
	}
}

static 
void
acollect(const char *as)
{
	int ogargc = gargc;

	gpathp = gpath; *gpathp = 0; globbed = 0;
	expand(as);
	if (gargc != ogargc)
		sort();
}

static 
void
sort(void)
{
	entry *p1, *p2, c;
	entry *Gvp = &gargv[gargc];

	p1 = sortbas;
	while (p1 < Gvp-1) {
		p2 = p1;
		while (++p2 < Gvp)
			if (strcmp(p1->text, p2->text) > 0)
				c = *p1, *p1 = *p2, *p2 = c;
		p1++;
	}
	sortbas = Gvp;
}

static 
void
expand(const char *as)
{
	const char *cs;
	const char *oldcs;
	char *sgpathp;
	struct stat stb;

	sgpathp = gpathp;
	cs = as;
	if (*cs == '~' && gpathp == gpath) {
		addpath('~');
		for (cs++; letter(*cs) || digit(*cs) || *cs == '-';)
			addpath(*cs++);
		if (!*cs || *cs == '/') {
			if (gpathp != gpath + 1) {
				*gpathp = 0;
				if (gethdir(gpath + 1))
					globerr = "Unknown user name after ~";
				(void) strcpy(gpath, gpath + 1);
			} 
			else {
				(void) strcpy(gpath, home);
			}
			gpathp = strend(gpath);
		}
	}
	while (!any(*cs, globchars)) {
		if (*cs == 0) {
			if (!globbed)
				Gcat(gpath, "");
			else if (stat(gpath, &stb) >= 0) {
				Gcat(gpath, "");
				globcnt++;
			}
			goto endit;
		}
		addpath(*cs++);
	}
	oldcs = cs;
	while (cs > as && *cs != '/')
		cs--, gpathp--;
	if (*cs == '/')
		cs++, gpathp++;
	*gpathp = 0;
	if (*oldcs == '{') {
		(void) execbrc(cs, ((char *)0));
		return;
	}
	matchdir(cs);
endit:
	gpathp = sgpathp;
	*gpathp = 0;
}

static 
void
matchdir(const char *pattern)
{
	struct stat stb;
	register struct dirent *dp;
	DIR *dirp;

#if 0
#ifdef	__linux__
	if (gpath == NULL || *gpath == '\0')
		gpath = "./";
#endif
#endif
	dirp = opendir((!gpath || !*gpath) ? "./" : gpath);
	if (dirp == NULL) {
		if (globbed)
			return;
		goto patherr2;
	}
	if (fstat(dirfd(dirp), &stb) < 0)
		goto patherr1;
	if (!isdir(stb)) {
		errno = ENOTDIR;
		goto patherr1;
	}
	while ((dp = readdir(dirp)) != NULL) {
		if (dp->d_ino == 0)
			continue;
		if (match(dp->d_name, pattern)) {
			Gcat(gpath, dp->d_name);
			globcnt++;
		}
	}
	closedir(dirp);
	return;

patherr1:
	closedir(dirp);
patherr2:
	globerr = "Bad directory components";
}

static 
int
execbrc(const char *p, const char *s)
{
	char restbuf[BUFSIZ + 2];
	const char *pe, *pm, *pl;
	int brclev = 0;
	char *lm, *sgpathp;

	for (lm = restbuf; *p != '{'; *lm++ = *p++)
		continue;
	for (pe = ++p; *pe; pe++)
	switch (*pe) {

	case '{':
		brclev++;
		continue;

	case '}':
		if (brclev == 0)
			goto pend;
		brclev--;
		continue;

	case '[':
		for (pe++; *pe && *pe != ']'; pe++)
			continue;
		continue;
	}
pend:
	brclev = 0;
	for (pl = pm = p; pm <= pe; pm++)
	switch (*pm & (QUOTE|TRIM)) {

	case '{':
		brclev++;
		continue;

	case '}':
		if (brclev) {
			brclev--;
			continue;
		}
		goto doit;

	case ','|QUOTE:
	case ',':
		if (brclev)
			continue;
doit:
#if 0
		savec = *pm;
		*pm = 0;
		strcpy(lm, pl);
		*pm = savec;
#else
		strncpy(lm, pl, pm-pl);
		lm[pm-pl] = 0;
#endif
		(void) strcat(restbuf, pe + 1);
		if (s == 0) {
			sgpathp = gpathp;
			expand(restbuf);
			gpathp = sgpathp;
			*gpathp = 0;
		} 
		else if (amatch(s, restbuf)) {
			return (1);
		}
		sort();
		pl = pm + 1;
		if (brclev)
			return (0);
		continue;

	case '[':
		for (pm++; *pm && *pm != ']'; pm++)
			continue;
		if (!*pm)
			pm--;
		continue;
	}
	if (brclev)
		goto doit;
	return (0);
}

static 
int
match(const char *s, const char *p)
{
	int c;
	const char *sentp;
	char sglobbed = globbed;

	if (*s == '.' && *p != '.')
		return (0);
	sentp = entp;
	entp = s;
	c = amatch(s, p);
	entp = sentp;
	globbed = sglobbed;
	return (c);
}

static 
int
amatch(const char *s, const char *p)
{
	register int scc;
	int ok, lc;
	char *sgpathp;
	struct stat stb;
	int c, cc;

	globbed = 1;
	for (;;) {
		scc = *s++ & TRIM;
		switch (c = *p++) {

		case '{':
			return (execbrc(p - 1, s - 1));

		case '[':
			ok = 0;
			lc = 077777;
			while ((cc = *p++) != 0) {
				if (cc == ']') {
					if (ok)
						break;
					return (0);
				}
				if (cc == '-') {
					if (lc <= scc && scc <= *p++)
						ok++;
				} else
					if (scc == (lc = cc))
						ok++;
			}
			if (cc == 0)
				if (ok)
					p--;
				else
					return 0;
			continue;

		case '*':
			if (!*p)
				return (1);
			if (*p == '/') {
				p++;
				goto slash;
			}
			s--;
			do {
				if (amatch(s, p))
					return (1);
			} while (*s++);
			return (0);

		case 0:
			return (scc == 0);

		default:
			if (c != scc)
				return (0);
			continue;

		case '?':
			if (scc == 0)
				return (0);
			continue;

		case '/':
			if (scc)
				return (0);
slash:
			s = entp;
			sgpathp = gpathp;
			while (*s)
				addpath(*s++);
			addpath('/');
			if (stat(gpath, &stb) == 0 && isdir(stb))
				if (*p == 0) {
					Gcat(gpath, "");
					globcnt++;
				} else
					expand(p);
			gpathp = sgpathp;
			*gpathp = 0;
			return (0);
		}
	}
}

static 
int
Gmatch(const char *s, const char *p)
{
	register int scc;
	int ok, lc;
	int c, cc;

	for (;;) {
		scc = *s++ & TRIM;
		switch (c = *p++) {

		case '[':
			ok = 0;
			lc = 077777;
			while ((cc = *p++) != 0) {
				if (cc == ']') {
					if (ok)
						break;
					return (0);
				}
				if (cc == '-') {
					if (lc <= scc && scc <= *p++)
						ok++;
				} else
					if (scc == (lc = cc))
						ok++;
			}
			if (cc == 0)
				if (ok)
					p--;
				else
					return 0;
			continue;

		case '*':
			if (!*p)
				return (1);
			for (s--; *s; s++)
				if (Gmatch(s, p))
					return (1);
			return (0);

		case 0:
			return (scc == 0);

		default:
			if ((c & TRIM) != scc)
				return (0);
			continue;

		case '?':
			if (scc == 0)
				return (0);
			continue;

		}
	}
}

static 
void
Gcat(const char *s1, const char *s2)
{
	int len = strlen(s1) + strlen(s2) + 1;

	if (len >= gnleft || gargc >= GAVSIZ - 1) {
		globerr = "Arguments too long";
	}
	else {
		gargc++;
		gnleft -= len;
		gargv[gargc].text = NULL;
		gargv[gargc - 1].text = strspl(s1, s2);
	}
}

static 
void
addpath(char c)
{

	if (gpathp >= lastgpathp)
		globerr = "Pathname too long";
	else {
		*gpathp++ = c;
		*gpathp = 0;
	}
}

static void
rscan(centry *t, int (*f)(char))
{
	const char *p;
	char c;

	while ((p = (t++)->text) != NULL) {
		if (f == tglob)
			if (*p == '~')
				gflag |= 2;
			else if (eq(p, "{") || eq(p, "{}"))
				continue;
		while ((c = *p++) != 0)
			(*f)(c);
	}
}

static 
int
tglob(char c)
{

	if (any(c, globchars))
		gflag |= c == '{' ? 2 : 1;
	return (c);
}

static int
letter(char c)
{
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}

static int
digit(char c)
{
	return (c >= '0' && c <= '9');
}

static 
int
any(int c, const char *s)
{
	while (*s) if (*s++ == c) return 1;
	return 0;
}

static 
int
cblklen(const centry *av)
{
	int i = 0;
	while ((av++)->text) i++;
	return i;
}

static 
int
blklen(const entry *av)
{
	int i = 0;
	while ((av++)->text) i++;
	return i;
}

void
blkfree(char **av)
{
	int i;
	for (i=0; av[i]; i++) free(av[i]);
}

static
void 
efree(entry *av) 
{
    int i;
    for (i=0; av[i].text; i++) free(av[i].text);
}

static
char *
strspl(const char *cp, const char *dp)
{
	char *ep = malloc(strlen(cp) + strlen(dp) + 1);
	if (ep == NULL)	fatal("Out of memory");

	strcpy(ep, cp);
	strcat(ep, dp);
	return ep;
}

static 
char **
copyblk(entry *v)
{
	int i;
	char **nv = malloc((blklen(v) + 1) * sizeof(char **));
	if (nv == NULL) fatal("Out of memory");

	for (i=0; v[i].text; i++) {
	    nv[i] = v[i].text;
	    v[i].text = NULL;
	}
	nv[i] = NULL;

	return nv;
}

static
char **
cloneblk(const centry *v)
{
	int i;
	char **nv = malloc((cblklen(v) + 1) * sizeof(char **));
	if (nv == NULL) fatal("Out of memory");

	for (i=0; v[i].text; i++) {
	    nv[i] = strdup(v[i].text);
	}
	nv[i] = NULL;

	return nv;
}

static
char *
strend(char *cp)
{
	while (*cp)
		cp++;
	return (cp);
}
/*
 * Extract a home directory from the password file
 * The argument points to a buffer where the name of the
 * user whose home directory is sought is currently.
 * We write the home directory of the user back there.
 */
static 
int
gethdir(char *homedir)
{
	register struct passwd *pp = getpwnam(homedir);

	if (!pp || homedir + strlen(pp->pw_dir) >= lastgpathp)
		return 1;
	strcpy(homedir, pp->pw_dir);
	return 0;
}


ftp/cmdtab.c
/*
 * Copyright (c) 1985, 1989 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * from: @(#)cmdtab.c	5.10 (Berkeley) 6/1/90
 */
char cmdtab_rcsid[] = 
  "$Id: cmdtab.c,v 1.6 1997/04/05 22:26:22 dholland Exp $";

#include <string.h>   /* for NULL */
#include "ftp_var.h"
#include "cmds.h"

/*
 * User FTP -- Command Tables.
 */

const char accounthelp[] = "send account command to remote server";
const char appendhelp[] =  "append to a file";
const char asciihelp[] =   "set ascii transfer type";
const char beephelp[] =    "beep when command completed";
const char binaryhelp[] =  "set binary transfer type";
const char casehelp[] =    "toggle mget upper/lower case id mapping";
const char cdhelp[] =      "change remote working directory";
const char cduphelp[] = "change remote working directory to parent directory";
const char chmodhelp[] =   "change file permissions of remote file";
const char connecthelp[] = "connect to remote tftp";
const char crhelp[] =      "toggle carriage return stripping on ascii gets";
const char deletehelp[] =  "delete remote file";
const char debughelp[] =   "toggle/set debugging mode";
const char dirhelp[] =     "list contents of remote directory";
const char disconhelp[] =  "terminate ftp session";
const char domachelp[] =   "execute macro";
const char formhelp[] =	"set file transfer format";
const char globhelp[] =	"toggle metacharacter expansion of local file names";
const char hashhelp[] =	"toggle printing `#' for each buffer transferred";
const char helphelp[] =	"print local help information";
const char idlehelp[] =	"get (set) idle timer on remote side";
const char lcdhelp[] =	"change local working directory";
const char lshelp[] =	"list contents of remote directory";
const char macdefhelp[] =  "define a macro";
const char mdeletehelp[] = "delete multiple files";
const char mdirhelp[] =    "list contents of multiple remote directories";
const char mgethelp[] =    "get multiple files";
const char mkdirhelp[] =   "make directory on the remote machine";
const char mlshelp[] =     "list contents of multiple remote directories";
const char modtimehelp[] = "show last modification time of remote file";
const char modehelp[] =    "set file transfer mode";
const char mputhelp[] =    "send multiple files";
const char newerhelp[] =   "get file if remote file is newer than local file ";
const char nlisthelp[] =   "nlist contents of remote directory";
const char nmaphelp[] =    "set templates for default file name mapping";
const char ntranshelp[] ="set translation table for default file name mapping";
const char passivehelp[] = "enter passive transfer mode";
const char porthelp[] =	   "toggle use of PORT cmd for each data connection";
const char prompthelp[] =  "force interactive prompting on multiple commands";
const char proxyhelp[] =   "issue command on alternate connection";
const char pwdhelp[] =     "print working directory on remote machine";
const char quithelp[] =    "terminate ftp session and exit";
const char quotehelp[] =   "send arbitrary ftp command";
const char receivehelp[] = "receive file";
const char regethelp[] =   "get file restarting at end of local file";
const char remotehelp[] =  "get help from remote server";
const char renamehelp[] =  "rename file";
const char restarthelp[] = "restart file transfer at bytecount";
const char rmdirhelp[] =   "remove directory on the remote machine";
const char rmtstatushelp[]="show status of remote machine";
const char runiquehelp[] = "toggle store unique for local files";
const char resethelp[] =   "clear queued command replies";
const char sendhelp[] =    "send one file";
const char sitehelp[] =    "send site specific command to remote server\n"
	    "\t\tTry \"rhelp site\" or \"site help\" for more information";
const char shellhelp[] =   "escape to the shell";
const char sizecmdhelp[] = "show size of remote file";
const char statushelp[] =  "show current status";
const char structhelp[] =  "set file transfer structure";
const char suniquehelp[] = "toggle store unique on remote machine";
const char systemhelp[] =  "show remote system type";
const char tenexhelp[] =   "set tenex file transfer type";
const char tickhelp[] =    "toggle printing byte counter during transfers";
const char tracehelp[] =   "toggle packet tracing";
const char typehelp[] =    "set file transfer type";
const char umaskhelp[] =   "get (set) umask on remote side";
const char userhelp[] =    "send new user information";
const char verbosehelp[] = "toggle verbose mode";

struct cmd cmdtab[] = {
	{ "!",		shellhelp,	0, 0, 0, NULL, NULL, shell },
	{ "$",		domachelp,	1, 0, 0, domacro, NULL, NULL },
	{ "account",	accounthelp,	0, 1, 1, account, NULL, NULL },
	{ "append",	appendhelp,	1, 1, 1, put, NULL, NULL },
	{ "ascii",	asciihelp,	0, 1, 1, NULL, setascii, NULL },
	{ "bell",	beephelp,	0, 0, 0, NULL, setbell, NULL },
	{ "binary",	binaryhelp,	0, 1, 1, NULL, setbinary, NULL },
	{ "bye",	quithelp,	0, 0, 0, NULL, quit, NULL },
	{ "case",	casehelp,	0, 0, 1, NULL, setcase, NULL },
	{ "cd",		cdhelp,		0, 1, 1, cd, NULL, NULL },
	{ "cdup",	cduphelp,	0, 1, 1, NULL, cdup, NULL },
	{ "chmod",	chmodhelp,	0, 1, 1, do_chmod, NULL, NULL },
	{ "close",	disconhelp,	0, 1, 1, NULL, disconnect, NULL },
	{ "cr",		crhelp,		0, 0, 0, NULL, setcr, NULL },
	{ "delete",	deletehelp,	0, 1, 1, delete_cmd, NULL, NULL },
	{ "debug",	debughelp,	0, 0, 0, setdebug, NULL, NULL },
	{ "dir",	dirhelp,	1, 1, 1, ls, NULL, NULL },
	{ "disconnect",	disconhelp,	0, 1, 1, NULL, disconnect, NULL },
	{ "exit",	quithelp,	0, 0, 0, NULL, quit, NULL },
	{ "form",	formhelp,	0, 1, 1, NULL, setform, NULL },
	{ "get",	receivehelp,	1, 1, 1, get, NULL, NULL },
	{ "glob",	globhelp,	0, 0, 0, NULL, setglob, NULL },
	{ "hash",	hashhelp,	0, 0, 0, NULL, sethash, NULL },
	{ "help",	helphelp,	0, 0, 1, help, NULL, NULL },
	{ "idle",	idlehelp,	0, 1, 1, idle_cmd, NULL, NULL },
	{ "image",	binaryhelp,	0, 1, 1, NULL, setbinary, NULL },
	{ "lcd",	lcdhelp,	0, 0, 0, lcd, NULL, NULL },
	{ "ls",		lshelp,		1, 1, 1, ls, NULL, NULL },
	{ "macdef",	macdefhelp,	0, 0, 0, macdef, NULL, NULL },
	{ "mdelete",	mdeletehelp,	1, 1, 1, mdelete, NULL, NULL },
	{ "mdir",	mdirhelp,	1, 1, 1, mls, NULL, NULL },
	{ "mget",	mgethelp,	1, 1, 1, mget, NULL, NULL },
	{ "mkdir",	mkdirhelp,	0, 1, 1, makedir, NULL, NULL },
	{ "mls",	mlshelp,	1, 1, 1, mls, NULL, NULL },
	{ "mode",	modehelp,	0, 1, 1, NULL, setmode, NULL },
	{ "modtime",	modtimehelp,	0, 1, 1, modtime, NULL, NULL },
	{ "mput",	mputhelp,	1, 1, 1, mput, NULL, NULL },
	{ "newer",	newerhelp,	1, 1, 1, newer, NULL, NULL },
	{ "nmap",	nmaphelp,	0, 0, 1, setnmap, NULL, NULL },
	{ "nlist",	nlisthelp,	1, 1, 1, ls, NULL, NULL },
	{ "ntrans",	ntranshelp,	0, 0, 1, setntrans, NULL, NULL },
	{ "open",	connecthelp,	0, 0, 1, setpeer, NULL, NULL },
	{ "prompt",	prompthelp,	0, 0, 0, NULL, setprompt, NULL },
        { "passive",    passivehelp,    0, 0, 0, NULL, setpassive, NULL },
	{ "proxy",	proxyhelp,	0, 0, 1, doproxy, NULL, NULL },
	{ "sendport",	porthelp,	0, 0, 0, NULL, setport, NULL },
	{ "put",	sendhelp,	1, 1, 1, put, NULL, NULL },
	{ "pwd",	pwdhelp,	0, 1, 1, NULL, pwd, NULL },
	{ "quit",	quithelp,	0, 0, 0, NULL, quit, NULL },
	{ "quote",	quotehelp,	1, 1, 1, quote, NULL, NULL },
	{ "recv",	receivehelp,	1, 1, 1, get, NULL, NULL },
	{ "reget",	regethelp,	1, 1, 1, reget, NULL, NULL },
	{ "rstatus",	rmtstatushelp,	0, 1, 1, rmtstatus, NULL, NULL },
	{ "rhelp",	remotehelp,	0, 1, 1, rmthelp, NULL, NULL },
	{ "rename",	renamehelp,	0, 1, 1, renamefile, NULL, NULL },
	{ "reset",	resethelp,	0, 1, 1, NULL, reset, NULL },
	{ "restart",	restarthelp,	1, 1, 1, restart, NULL, NULL },
	{ "rmdir",	rmdirhelp,	0, 1, 1, removedir, NULL, NULL },
	{ "runique",	runiquehelp,	0, 0, 1, NULL, setrunique, NULL },
	{ "send",	sendhelp,	1, 1, 1, put, NULL, NULL },
	{ "site",	sitehelp,	0, 1, 1, site, NULL, NULL },
	{ "size",	sizecmdhelp,	1, 1, 1, sizecmd, NULL, NULL },
	{ "status",	statushelp,	0, 0, 1, NULL, status, NULL },
	{ "struct",	structhelp,	0, 1, 1, NULL, setstruct, NULL },
	{ "system",	systemhelp,	0, 1, 1, NULL, syst, NULL },
	{ "sunique",	suniquehelp,	0, 0, 1, NULL, setsunique, NULL },
	{ "tenex",	tenexhelp,	0, 1, 1, NULL, settenex, NULL },
	{ "tick",	tickhelp,	0, 0, 0, NULL, settick, NULL },
	{ "trace",	tracehelp,	0, 0, 0, NULL, settrace, NULL },
	{ "type",	typehelp,	0, 1, 1, settype, NULL, NULL },
	{ "user",	userhelp,	0, 1, 1, user, NULL, NULL },
	{ "umask",	umaskhelp,	0, 1, 1, do_umask, NULL, NULL },
	{ "verbose",	verbosehelp,	0, 0, 0, NULL, setverbose, NULL },
	{ "?",		helphelp,	0, 0, 1, help, NULL, NULL },
	{ 0 },
};

int	NCMDS = (sizeof (cmdtab) / sizeof (cmdtab[0])) - 1;
